libmisc/failure.c

```diff
@@ -86,12 +86,18 @@ void failure (uid_t uid, const char *tty, struct faillog *fl)
 	 */
 
 	if (   (lseek (fd, offset_uid, SEEK_SET) != offset_uid)
-	    || (write (fd, (char *) fl, sizeof *fl) != (ssize_t) sizeof *fl)
-	    || (close (fd) != 0)) {
+	    || (write (fd, (char *) fl, sizeof *fl) != (ssize_t) sizeof *fl)) {
 		SYSLOG ((LOG_WARN,
 		         "Can't write faillog entry for UID %lu in %s.",
 		         (unsigned long) uid, FAILLOG_FILE));
 		(void) close (fd);
+		return;
+	}
+
+	if (close (fd) != 0) {
+		SYSLOG ((LOG_WARN,
+		         "Can't write faillog entry for UID %lu in %s.",
+		         (unsigned long) uid, FAILLOG_FILE));
 	}
 }
```

Comment:
A few things can be improved here:

-  The cast to `(const void *)` seems useless.  _Every_ pointer converts automatically to it.

-  I suggest reducing nested logic with the following `goto`s.

```c
	if (write (fd, failent, sizeof *failent) != (ssize_t) sizeof *failent)
		goto write_err;

	if (close (fd) != 0)
		goto close_err;

	return;

write_err:
	(void) close (fd);
close_err:
	SYSLOG ((LOG_WARN, ...));
```

Reply:
I personally don't like introducing `goto` statements for these cases: they decouple the error handling from the cause and are sensitive in their order, e.g. in your suggestion `write_err` might override the `errno` printed in `close_err`.

@hallyn @ikerexxe what is your preferred style for this?